File: firebase/firestore/node/admins.dart
// @dart=2.9

import "package:firestore/helpers.dart";
import "package:firestore/services.dart";

Future<Map<String, List<String>>> getAdmins() async => {
	await for(final List<String> row in csvReadLines(DataFiles.admins))
		row [0]: row.sublist(1)
};

Future<void> setClaims(Map<String, List<String>> admins) async {
	for (final MapEntry<String, List<String>> entry in admins.entries) {
		final String email = entry.key;
		final List<String> scopes = entry.value;
		if (!scopes.every(Scopes.scopes.contains)) {
			throw ArgumentError.value(
				scopes.toString(), 
				"admin scopes", 
				"Unrecognized scopes for $email"
			);
		}
		Logger.verbose("Setting claims for $email");
		if (entry.value.isEmpty) {
			Logger.warning("Removing admin privileges for $email");
		}
		await Logger.logValue<Map<String, dynamic>>(
			"Previous claims for $email", () => Auth.getClaims(email)
		);
		await Auth.setScopes(email, scopes);
	}
}

Future<void> main() async {
	Args.initLogger("Setting up admins...");

	// Log the value whether or not --upload is passed.
	final Map<String, List<String>> admins = 
		await Logger.logValue("admins", getAdmins);

	if (Args.upload) {
		await setClaims(admins);
	} else {
		Logger.warning("Did not upload admin claims. Use the --upload flag.");
	}

	await app.delete();
	Logger.info("Finished setting up admins");
}

File: firebase/firestore-py/bin/admins.py
from firebase_admin import delete_app
import csv

import lib.utils as utils
import lib.services as firebase

def get_admins(): 
	with open(utils.dir.admins) as file: return {
		row[0]: [i for i in row[1:] if i]
		for row in csv.reader(file)
	}

def set_claims(admins): 
	for email, scopes in admins.items(): 
		if not all(scope in firebase.SCOPES.union({""}) for scope in scopes): 
			raise ValueError(f"Unrecognized scopes for {email}: {scopes}")
		utils.logger.verbose(f"Setting claims for {email}")
		if not scopes: 
			utils.logger.warning(f"Removing admin priveleges for {email}")
		if utils.args.debug:
			utils.logger.debug(f"  Claims for {email}", firebase.get_claims(email))
		if utils.args.should_upload: firebase.set_scopes(email, scopes)
	if not utils.args.should_upload: 
		utils.logger.warning("Did not upload admin claims. Use the --upload flag.")

if __name__ == '__main__':
	utils.logger.info("Setting up admins...")
	admins = utils.logger.log_value("admins", get_admins)
	set_claims(admins)

	delete_app(firebase.app)
	utils.logger.info("Finished setting up admins")

